sdktests/server_side_persistence_read_write.go

```diff
@@ -0,0 +1,469 @@
+		"flag-key":          basicFlagValidationMatcher("flag-key", 200, "new-value"),
+		"uncached-flag-key": basicFlagValidationMatcher("uncached-flag-key", 100, "value"),
+	})
+}
+
+func (s *ServerSidePersistentTests) dataSourceDeletesRespectVersioning(t *ldtest.T) {
+	require.NoError(t, s.persistentStore.Reset())
+
+	persistence := NewPersistence()
+	persistence.SetStore(servicedef.SDKConfigPersistentStore{
+		Type: s.persistentStore.Type(),
+		DSN:  s.persistentStore.DSN(),
+	})
+	persistence.SetCache(servicedef.SDKConfigPersistentCache{
+		Mode: servicedef.Off,
+	})
+
+	sdkData := s.makeSDKDataWithFlag("flag-key", 100, ldvalue.String("value"))
+	stream, configurers := s.setupDataSources(t, sdkData)
+	configurers = append(configurers, persistence)
+
+	_ = NewSDKClient(t, s.baseSDKConfigurationPlus(configurers...)...)
+	s.eventuallyRequireDataStoreInit(t, "launchdarkly")
+
+	require.NoError(t, s.persistentStore.WriteData("launchdarkly:features", s.initialFlags))
+
+	// Lower versioned deletes are ignored
+	stream.StreamingService().PushDelete("flags", "flag-key", 1)
+	s.neverValidateFlagData(t, "launchdarkly", map[string]m.Matcher{
+		"flag-key":          basicDeletedFlagValidationMatcher(1),
+		"uncached-flag-key": basicFlagValidationMatcher("uncached-flag-key", 100, "fallthrough"),
+	})
+
+	// Higher versioned deletes are applied
+	stream.StreamingService().PushDelete("flags", "flag-key", 200)
+	s.eventuallyValidateFlagData(t, "launchdarkly", map[string]m.Matcher{
+		"flag-key":          basicDeletedFlagValidationMatcher(200),
+		"uncached-flag-key": basicFlagValidationMatcher("uncached-flag-key", 100, "fallthrough"),
+	})
+}
+
+func (s *ServerSidePersistentTests) ignoresDirectDatabaseModifications(t *ldtest.T, cacheConfig servicedef.SDKConfigPersistentCache) {
+	require.NoError(t, s.persistentStore.Reset())
+
+	persistence := NewPersistence()
+	persistence.SetStore(servicedef.SDKConfigPersistentStore{
+		Type: s.persistentStore.Type(),
+		DSN:  s.persistentStore.DSN(),
+	})
+	persistence.SetCache(cacheConfig)
+
+	sdkData := s.makeSDKDataWithFlag("flag-key", 1, ldvalue.String("value"))
+	_, configurers := s.setupDataSources(t, sdkData)
+	configurers = append(configurers, persistence)
+
+	client := NewSDKClient(t, s.baseSDKConfigurationPlus(configurers...)...)
+	context := ldcontext.New("user-key")
+	s.eventuallyRequireDataStoreInit(t, "launchdarkly")
+
+	pollUntilFlagValueUpdated(t, client, "flag-key", context,
+		ldvalue.String("default"), ldvalue.String("value"), ldvalue.String("default"))
+
+	require.NoError(t, s.persistentStore.WriteData("launchdarkly:features", s.initialFlags))
+
+	// This key was already cached, so it shouldn't see the change above.
+	h.RequireNever(t,
+		checkForUpdatedValue(t, client, "flag-key", context,
+			ldvalue.String("value"), ldvalue.String("new-value"), ldvalue.String("default")),
+		time.Millisecond*500, time.Millisecond*20, "flag-key was incorrectly updated")
```

Comment:
These time-based assertions must depend on the TTL value to be correct right? Should they be derived from it, like cacheConfig.TTL + X milliseconds or something?